include/bio/plain_io/misc.hpp

```diff
@@ -0,0 +1,122 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2021, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2021, Knut Reinert & MPI für molekulare Genetik
+// Copyright (c) 2020-2021, deCODE Genetics
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/bio/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides utility types for bio::plain_io::reader and bio::plain_io::writer.
+ * \author Hannes Hauswedell <hannes.hauswedell AT decode.is>
+ */
+
+#pragma once
+
+#include <stdexcept>
+#include <string_view>
+#include <vector>
+
+#include <bio/platform.hpp>
+
+namespace bio::plain_io
+{
+
+/*!\addtogroup plain_io
+ * \{
+ */
+
+//!\brief The value type of bio::plaintext_file_input if every line is split into fields.
+struct record
+{
+    //!\brief The entire line (exluding EOL characters but including delimiters).
+    std::string_view              line;
+    //!\brief A range of the individual fields (without delimiters or EOL characters).
+    std::vector<std::string_view> fields;
+};
+
+//!\brief Option to switch between reading-by-line and splitting a line into fields.
+enum class record_kind
+{
+    line,           //!< Only the line is provided.
+    line_and_fields //!< The line is provided and also individual fields (bio::plaintext_record).
+};
+
+/*!\brief A helper for specifying the header of a bio::plaintext_file_input.
+ * \tparam record_kind Whether to split lines on delimiter (e.g. TSV files) or not.
+ *
+ * \details
+ *
+ * TODO explain how this can be used easily
+ */
+class header_kind
+{
+private:
+    //!\brief The state of the variable, encoded as the value of the char or the special values -200 and -300.
+    int state = -200;
+
+public:
+    //!\brief The state representing "no header".
+    static constexpr struct
+    {
+    } none{};
+    //!\brief The state representing "first line is header".
+    static constexpr struct
+    {
+    } first_line{};
+
+    //!\brief Type of the state representing "all lines that start with character X".
+    struct starts_with
+    {
+        //!\privatesection
+        //!\brief Store the chracter.
+        char c;
+    };
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr header_kind() noexcept                    = default;             //!< Defaulted.
+    constexpr header_kind(header_kind const &) noexcept = default;             //!< Defaulted.
+    constexpr header_kind(header_kind &&) noexcept      = default;             //!< Defaulted.
+    constexpr header_kind & operator=(header_kind const &) noexcept = default; //!< Defaulted.
+    constexpr header_kind & operator=(header_kind &&) noexcept = default;      //!< Defaulted.
```

Comment:
This is what clang format does.. You really think this is more pretty with the defaults not all starting at the same spot? :smile_cat:

Reply:
I like it :)